### PR TITLE
HIP fix compiling with Tuning

### PR DESCRIPTION
**Pull request by masterleinad**
We apparently didn't update this location in https://github.com/kokkos/kokkos/pull/3953 accordingly.

### CODE REVIEW

**Review by DavidPoliakoff — APPROVED** [review]

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp [hunk 1]
```diff
@@ -185,20 +185,18 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
...
     using closure_type =
         ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
                     Kokkos::Experimental::HIP>;
     unsigned block_size =
         Kokkos::Experimental::Impl::hip_get_max_blocksize<closure_type,
-                                                          LaunchBounds>(f);
-    if (block_size > 0) {
-      return block_size;
-    } else {
+                                                          LaunchBounds>();
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:194]: I don't think that's correct. You should add the instance instance.

@arghdos, you wrote that code do you agree that the instance is missing.

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp [hunk 2]
```diff
@@ -433,13 +431,13 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
   template <typename Policy, typename Functor>
-  static int max_tile_size_product(const Policy&, const Functor& f) {
+  static int max_tile_size_product(const Policy&, const Functor&) {
     using closure_type =
         ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>,
                        ReducerType, Kokkos::Experimental::HIP>;
     unsigned block_size =
         Kokkos::Experimental::Impl::hip_get_max_blocksize<closure_type,
-                                                          LaunchBounds>(f);
+                                                          LaunchBounds>();
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:440]: Same as above.

**Review by Rombur — CHANGES_REQUESTED** [review]

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:194]: This doesn't work. I get
```
[ 62%] Building CXX object core/unit_test/CMakeFiles/KokkosCore_IncrementalTest_HIP.dir/generated/Test06_ParallelFor_MDRangePolicy_HIP.cpp.o
In file included from /tmp/kokkos/build/core/unit_test/generated/Test06_ParallelFor_MDRangePolicy_HIP.cpp:53:
In file included from /tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:45:
In file included from /tmp/kokkos/core/src/Kokkos_Core.hpp:56:
In file included from /tmp/kokkos/build/KokkosCore_Config_DeclareBackend.hpp:48:
In file included from /tmp/kokkos/core/src/decl/Kokkos_Declare_HIP.hpp:49:
In file included from /tmp/kokkos/core/src/Kokkos_HIP.hpp:60:
root@lascaux01:/tmp/kokkos/build# head -100 errors
[ 50%] Built target kokkoscore
[ 50%] Built target kokkoscontainers
[ 62%] Built target kokkos_gtest
[ 62%] Building CXX object core/unit_test/CMakeFiles/KokkosCore_IncrementalTest_HIP.dir/generated/Test06_ParallelFor_MDRangePolicy_HIP.cpp.o
In file included from /tmp/kokkos/build/core/unit_test/generated/Test06_ParallelFor_MDRangePolicy_HIP.cpp:53:
In file included from /tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:45:
In file included from /tmp/kokkos/core/src/Kokkos_Core.hpp:56:
In file included from /tmp/kokkos/build/KokkosCore_Config_DeclareBackend.hpp:48:
In file included from /tmp/kokkos/core/src/decl/Kokkos_Declare_HIP.hpp:49:
In file included from /tmp/kokkos/core/src/Kokkos_HIP.hpp:60:
In file included from /tmp/kokkos/core/src/HIP/Kokkos_HIP_Parallel_Range.hpp:52:
/tmp/kokkos/core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp:148:28: error: no matching function for call to object of type 'const Test06::MDFunctor'
    unsigned total_shmem = f(block_size);
                           ^
/tmp/kokkos/core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp:269:10: note: in instantiation of function template specialization 'Kokkos::Experimental::Impl::hip_internal_get_block_size<Kokkos::Experimental::Impl::BlockType::Max, Kokkos::Impl::ParallelFor<Test06::MDFunctor, Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::Experimental::HIP>, Kokkos::LaunchBounds<0, 0>, Test06::MDFunctor>' requested here
  return hip_internal_get_block_size<BlockType::Max, DriverType, LaunchBounds>(
         ^
/tmp/kokkos/core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:193:37: note: in instantiation of function template specialization 'Kokkos::Experimental::Impl::hip_get_max_blocksize<Kokkos::Impl::ParallelFor<Test06::MDFunctor, Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::Experimental::HIP>, Kokkos::LaunchBounds<0, 0>, Test06::MDFunctor>' requested here
        Kokkos::Experimental::Impl::hip_get_max_blocksize<closure_type,
                                    ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:317:20: note: in instantiation of function template specialization 'Kokkos::Impl::ParallelFor<Test06::MDFunctor, Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::Experimental::HIP>::max_tile_size_product<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Test06::MDFunctor>' requested here
    return driver::max_tile_size_product(policy, functor);
                   ^
/tmp/kokkos/core/src/Kokkos_Tuners.hpp:588:14: note: in instantiation of function template specialization 'Kokkos::Tools::Impl::Impl::SimpleTeamSizeCalculator::get_mdrange_max_tile_size_product<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Test06::MDFunctor>' requested here
        calc.get_mdrange_max_tile_size_product(policy, functor, tag);
             ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:385:36: note: in instantiation of function template specialization 'Kokkos::Tools::Experimental::MDRangeTuner<2>::MDRangeTuner<Test06::MDFunctor, Kokkos::ParallelForTag, Kokkos::Tools::Impl::Impl::SimpleTeamSizeCalculator, Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>' requested here
        return (map.emplace(label, Tuner(label, policy, functor, tag,
                                   ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:456:3: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  generic_tune_policy<Experimental::MDRangeTuner<rank>>(
  ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:568:5: note: in instantiation of function template specialization 'Kokkos::Tools::Impl::tune_policy<Test06::MDFunctor, Kokkos::ParallelForTag, Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>' requested here
    tune_policy(context_id, label, policy, functor, Kokkos::ParallelForTag{});
    ^
/tmp/kokkos/core/src/Kokkos_Parallel.hpp:136:24: note: in instantiation of function template specialization 'Kokkos::Tools::Impl::begin_parallel_for<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, const Test06::MDFunctor>' requested here
  Kokkos::Tools::Impl::begin_parallel_for(inner_policy, functor, str, kpID);
                       ^
/tmp/kokkos/core/src/Kokkos_Parallel.hpp:173:13: note: in instantiation of function template specialization 'Kokkos::parallel_for<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Test06::MDFunctor>' requested here
  ::Kokkos::parallel_for(policy, functor, str);
            ^
/tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:168:13: note: in instantiation of function template specialization 'Kokkos::parallel_for<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Test06::MDFunctor>' requested here
    Kokkos::parallel_for("MDRange2D", mdPolicy_2D, Functor_2D);
            ^
/tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:252:8: note: in instantiation of member function 'Test06::TestMDRangePolicy<Kokkos::Experimental::HIP>::mdRange2D' requested here
  test.mdRange2D();
       ^
/tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:70:8: note: candidate function not viable: requires 2 arguments, but 1 was provided
  void operator()(const int i, const int j) const {
       ^
/tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:76:8: note: candidate function not viable: requires 3 arguments, but 1 was provided
  void operator()(const int i, const int j, const int k) const {
       ^
/tmp/kokkos/core/unit_test/incremental/Test06_ParallelFor_MDRangePolicy.hpp:82:8: note: candidate function not viable: requires 4 arguments, but 1 was provided
  void operator()(const int i, const int j, const int k, const int l) const {
       ^
```
all over the place. How is `ParallelFor` using LDS?

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:440]:
```
In file included from /tmp/kokkos/build/core/unit_test/generated/Test14_MDRangeReduce_HIP.cpp:53:
In file included from /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:49:
In file included from /tmp/kokkos/core/src/Kokkos_Core.hpp:56:
In file included from /tmp/kokkos/build/KokkosCore_Config_DeclareBackend.hpp:48:
In file included from /tmp/kokkos/core/src/decl/Kokkos_Declare_HIP.hpp:49:
In file included from /tmp/kokkos/core/src/Kokkos_HIP.hpp:60:
In file included from /tmp/kokkos/core/src/HIP/Kokkos_HIP_Parallel_Range.hpp:52:
/tmp/kokkos/core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp:148:28: error: no matching function for call to object of type 'const (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9)'
    unsigned total_shmem = f(block_size);
                           ^
/tmp/kokkos/core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp:269:10: note: in instantiation of function template specialization 'Kokkos::Experimental::Impl::hip_internal_get_block_size<Kokkos::Experimental::Impl::BlockType::Max, Kokkos::Impl::ParallelReduce<(lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::InvalidType, Kokkos::Experimental::HIP>, Kokkos::LaunchBounds<0, 0>, (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9)>' requested here
  return hip_internal_get_block_size<BlockType::Max, DriverType, LaunchBounds>(
         ^
/tmp/kokkos/core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:439:37: note: in instantiation of function template specialization 'Kokkos::Experimental::Impl::hip_get_max_blocksize<Kokkos::Impl::ParallelReduce<(lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::InvalidType, Kokkos::Experimental::HIP>, Kokkos::LaunchBounds<0, 0>, (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9)>' requested here
        Kokkos::Experimental::Impl::hip_get_max_blocksize<closure_type,
                                    ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:327:20: note: in instantiation of function template specialization 'Kokkos::Impl::ParallelReduce<(lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::InvalidType, Kokkos::Experimental::HIP>::max_tile_size_product<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9)>' requested here
    return driver::max_tile_size_product(policy, functor);
                   ^
/tmp/kokkos/core/src/Kokkos_Tuners.hpp:588:14: note: in instantiation of function template specialization 'Kokkos::Tools::Impl::Impl::SimpleTeamSizeCalculator::get_mdrange_max_tile_size_product<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9)>' requested here
        calc.get_mdrange_max_tile_size_product(policy, functor, tag);
             ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:385:36: note: in instantiation of function template specialization 'Kokkos::Tools::Experimental::MDRangeTuner<2>::MDRangeTuner<(lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::ParallelReduceTag, Kokkos::Tools::Impl::Impl::SimpleTeamSizeCalculator, Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>' requested here
        return (map.emplace(label, Tuner(label, policy, functor, tag,
                                   ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:456:3: note: (skipping 2 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  generic_tune_policy<Experimental::MDRangeTuner<rank>>(
  ^
/tmp/kokkos/core/src/impl/Kokkos_Profiling.hpp:647:35: note: in instantiation of function template specialization 'Kokkos::Tools::Impl::ReductionSwitcher<Kokkos::InvalidType>::tune<(lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), Kokkos::ParallelReduceTag, Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>>' requested here
  ReductionSwitcher<ReducerType>::tune(context_id, label, policy, functor,
                                  ^
/tmp/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:860:26: note: in instantiation of function template specialization 'Kokkos::Tools::Impl::begin_parallel_reduce<Kokkos::InvalidType, Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, const (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9)>' requested here
    Kokkos::Tools::Impl::begin_parallel_reduce<
                         ^
/tmp/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:1020:69: note: in instantiation of member function 'Kokkos::Impl::ParallelReduceAdaptor<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), double>::execute' requested here
  Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(
                                                                    ^
/tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:123:13: note: in instantiation of function template specialization 'Kokkos::parallel_reduce<Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, (lambda at /tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9), double>' requested here
    Kokkos::parallel_reduce(
            ^
/tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:178:8: note: in instantiation of member function 'Test::TestMDRangeReduce<Kokkos::Experimental::HIP>::reduce_MDRange' requested here
  test.reduce_MDRange();
       ^
/tmp/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp:125:9: note: candidate function not viable: requires 3 arguments, but 1 was provided
        KOKKOS_LAMBDA(const int i, const int j, value_type& update_value) {
        ^
/tmp/kokkos/core/src/setup/Kokkos_Setup_HIP.hpp:56:23: note: expanded from macro 'KOKKOS_LAMBDA'
#define KOKKOS_LAMBDA [=] __host__ __device__
```

**Review by Rombur — APPROVED** [review]
Thanks @arghdos

**Review by dalg24 — APPROVED** [review]

**Pull request merged**